sdk/src/fast_int_math.rs

```diff
@@ -0,0 +1,131 @@
+/// return the bit width of the integer type
+pub const fn num_bits<T>() -> u32 {
+    (std::mem::size_of::<T>() * 8) as u32
+}
+
+/// generate fn to compute the floor of log2 for integer types
+macro_rules! gen_log2_floor {
+    ($n: ident, $t: ty) => {
+        pub fn $n(x: $t) -> $t {
+            (num_bits::<$t>() - x.leading_zeros() - 1) as $t
+        }
+    };
+}
+gen_log2_floor!(log2_floor_u32, u32);
+gen_log2_floor!(log2_floor_u64, u64);
+
+/// generate fn to compute the ceil of log2 for integer types
+macro_rules! gen_log2_ceil {
+    ($n: ident, $t: ty) => {
+        pub fn $n(x: $t) -> $t {
+            (num_bits::<$t>() - (x - 1 as $t).leading_zeros()) as $t
```

Comment:
I don't quite understand why do we have here `-1` which was preceding your changes.
shouldn't it be just `num_bits - cnt_leading_zeroes`?

Reply:
Let's say ceil(log2(2)) = 1.
for 32bit, 32-leading_zeros(2) = 32- 30 = 2
32 - leading_zeros(2-1) = 32 - 31 = 1